File: 56-merge-intervals/merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> result;
        result.push_back(intervals[0]);
        for (auto i = 1; i < intervals.size(); i++) {
            if (max(intervals[i][0], result[result.size() - 1][0]) 
                    <= min(intervals[i][1], result[result.size() - 1][1])) 
                result[result.size() - 1] = 
                    {min(intervals[i][0], result[result.size() - 1][0]), 
                        max(intervals[i][1], result[result.size() - 1][1])};
            else
                result.push_back(intervals[i]);
        }
        return result;
    }
};

File: 56-merge-intervals/merge-intervals.go
func merge(intervals [][]int) [][]int {
    slices.SortFunc(intervals,func(a, b []int) int {
        if a[0] == b[0] {
            return cmp.Compare(a[1], b[1])
        }
        return cmp.Compare(a[0], b[0])
    })

    res := [][]int{intervals[0]}

    for i := 1; i < len(intervals); i++ {
        if isMerging(res[len(res) - 1], intervals[i]) {
            res[len(res) - 1] = merged(res[len(res) - 1], intervals[i])
        } else {
            res = append(res, intervals[i])
        }
    }

    return res
}

func isMerging(interval1, interval2 []int) bool {
    return max(interval1[0], interval2[0]) <= min(interval1[1], interval2[1]) 
}

func merged(interval1, interval2 []int) []int {
    return []int{min(interval1[0], interval2[0]),max(interval1[1], interval2[1])}
}